### ISSUE TITLE
Resource shutdown ignores dependencies between resources

### ISSUE THREAD

**Issue by saulbein** [ets-labs/python-dependency-injector#432]
First of all, thank you for the great library.

I have encountered an issue with a bit of an unusual case of having dependencies between resources where both of them have cleanup, and the order of the cleanup matters. This is the container we are using:
```
class Container(containers.DeclarativeContainer):
    config = providers.Configuration()

    engine = providers.Singleton(...)

    session = providers.Resource(create_session, engine)
    token = providers.Resource(locked_token, session)

def create_session(engine):
    db_session = Session(engine)
    try:
        yield db_session
    finally:
        db_session.close()

def locked_token(session):
    token = session.query(...)
    session.commit()
    try:
        yield token
    finally:
        session.refresh(token)
        ...
        session.commit()
```

Resource initialization works completely fine, the session gets initialized before the token. But at shutdown of the application, we tried calling `container.shutdown()`, and what happened was that it closed the session first, thus making the token cleanup crash (since it got a new session - SQLAlchemy sessions create a new one on `close()`).

In other words, the change requested here is that the container shuts down resources in reverse order of starting them.

**Comment by rmk135** [comment]
Hey @saulbein ,

That's an interesting case. Thanks for opening the issue. Yep, that's something that would be cool to work out of the box. There could be a couple of ways it could be implemented. I'll think of it and get back in a while.

As a workaround I would suggest to manually shutdown resources in a proper order.

**Comment by saulbein** [comment]
Hi again @rmk135,

The workaround has been fine for a while now. 

We have a different framework that is responsible for application lifetime ([Prefect](https://www.prefect.io/)) and we had the container integrated with it's tasks, so that we can easily provide external dependencies. Since we have multiple projects using this setup, I would like to factor out the code linking these two frameworks, but I can't do that unless `container.shutdown()` respects the shutdown order - the linker code can't know about what the container has, just that it is a container.

Not trying to rush you or anything, just expanding on the use case.

**Comment by rmk135** [comment]
Got it @saulbein. Thanks for sharing the use case. I’m going to estimate the effort for the feature by the end of the week. Will let you know, when I have more updates

**Comment by rmk135** [comment]
Hey @saulbein. I have an implementation of ``container.shutdown_resources()`` that respects resource dependencies. This means that next code:

```python
"""Resources shutdown ordering example."""

import asyncio

from dependency_injector import containers, providers


async def async_resource(name, shutdown_delay, **dependencies):
    print(f'Init resource {name}')
    yield
    await asyncio.sleep(shutdown_delay)
    print(f'Shutdown resource {name}')


class Container(containers.DeclarativeContainer):

    resource_1 = providers.Resource(
        async_resource,
        name='resource 1',
        shutdown_delay=0.1,
    )

    resource_2 = providers.Resource(
        async_resource,
        name='resource 2',
        shutdown_delay=0.2,
        dependency=resource_1,
    )

    resource_3 = providers.Resource(
        async_resource,
        name='resource 3',
        shutdown_delay=0.3,
        dependency=resource_2,
    )


async def main():
    container = Container()
    await container.init_resources()
    await container.shutdown_resources()


if __name__ == '__main__':
    asyncio.run(main())
```

will always produce this output:

```
Init resource resource 1
Init resource resource 2
Init resource resource 3
Shutdown resource resource 1
Shutdown resource resource 2
Shutdown resource resource 3
```

I will need some more time for refactoring and writing more tests. Expect the release early next week.

**Comment by rmk135** [comment]
Hi @saulbein . I've released version `4.34.1`. It respects resources dependencies shutdown order. Thanks a lot for reporting the issue. Also please let me know if anything else pops up - I'll be glad to help.

**Comment by saulbein** [comment]
@rmk135 Shouldn't it be the reverse order?
```
Init resource resource 1
Init resource resource 2
Init resource resource 3
Shutdown resource resource 3
Shutdown resource resource 2
Shutdown resource resource 1
```
Or should I specify the order using the delays? If the shutdown of a resource takes a while, that might be an issue.

**Comment by rmk135** [comment]
> @rmk135 Shouldn't it be the reverse order?

It should. (facepalm) I've made completely wrong thing. Dummy mistake. Will work on the fix to reverse the order.

**Comment by rmk135** [comment]
Fixed in `4.34.2`

**Comment by saulbein** [comment]
No worries, we all make mistakes sometimes :)
Thanks for the fix!
